File: chapter-06-metadata/python/01-termstore/explore_term_store.py
#!/usr/bin/env python3
# explore_term_store.py
# Ejemplo: Explorar el Term Store con Microsoft Graph SDK para Python.
# El GraphServiceClient se inyecta por constructor (DI) desde el módulo común.

import asyncio
from msgraph import GraphServiceClient

from laexperiencia_sharepoint import get_graph_client


class TermStoreExplorer:
    def __init__(self, graph_client: GraphServiceClient):
        self.graph_client = graph_client

    async def explore(self):
        print("=" * 60)
        print("Explorando Term Store")
        print("=" * 60)
        print()

        try:
            # Obtener Term Store
            term_store = await self.graph_client.sites.by_site_id("root").term_store.get()

            print("📚 Term Store Info:")
            print(f"   ID: {term_store.id}")
            print(f"   Default Language: {term_store.default_language_tag}")
            print(f"   Languages: {', '.join(term_store.language_tags or [])}")
            print()

            # Listar grupos
            groups = await self.graph_client.sites.by_site_id("root").term_store.groups.get()

            print(f"📂 Grupos encontrados: {len(groups.value)}\n")

            for group in groups.value:
                print(f"   📁 {group.display_name}")
                print(f"      ID: {group.id}")
                print(f"      Descripción: {group.description or 'N/A'}")

                # Obtener conjuntos de términos
                sets = await self.graph_client.sites.by_site_id("root") \
                    .term_store.groups.by_group_id(group.id).sets.get()

                for set_item in sets.value:
                    set_name = set_item.localized_names[0].name if set_item.localized_names else set_item.id
                    print(f"         📚 {set_name}")

                print()

        except Exception as e:
            print(f"❌ Error: {e}")


async def main():
    # Cliente de Graph vía módulo común (client secret, app-only).
    graph_client = get_graph_client()
    explorer = TermStoreExplorer(graph_client)
    await explorer.explore()


if __name__ == "__main__":
    asyncio.run(main())

File: chapter-06-metadata/javascript/01-termstore/exploreTermStore.js
// exploreTermStore.js
// Example: Explore the Term Store with Microsoft Graph JavaScript SDK.
// The Graph client is injected via constructor (DI) from the common module.

const { getGraphClient } = require('la-experiencia-sharepoint-code/graphAuth');

class TermStoreExplorer {
    constructor(graphClient) {
        this.graphClient = graphClient;
    }

    async explore() {
        console.log('='.repeat(60));
        console.log('Explorando Term Store');
        console.log('='.repeat(60));
        console.log();

        try {
            // Get Term Store
            const termStore = await this.graphClient
                .api('/sites/root/termStore')
                .get();

            console.log('📚 Term Store Info:');
            console.log(`   ID: ${termStore.id}`);
            console.log(`   Default Language: ${termStore.defaultLanguage}`);
            console.log(`   Languages: ${termStore.languageTags?.join(', ') || 'N/A'}`);
            console.log();

            // List groups
            const groups = await this.graphClient
                .api('/sites/root/termStore/groups')
                .get();

            console.log(`📂 Grupos encontrados: ${groups.value.length}\n`);

            for (const group of groups.value) {
                console.log(`   📁 ${group.displayName}`);
                console.log(`      ID: ${group.id}`);
                console.log(`      Description: ${group.description || 'N/A'}`);

                // Get term sets
                const sets = await this.graphClient
                    .api(`/sites/root/termStore/groups/${group.id}/sets`)
                    .get();

                for (const set of sets.value) {
                    const setName = set.localizedNames?.[0]?.name || set.id;
                    console.log(`         📚 ${setName}`);
                }
                console.log();
            }

        } catch (error) {
            console.error('❌ Error:', error.message);
        }
    }
}

// Example usage: Graph client via common module (client secret, app-only).
async function main() {
    const graphClient = await getGraphClient();
    const explorer = new TermStoreExplorer(graphClient);
    await explorer.explore();
}

module.exports = { TermStoreExplorer };

// If run directly
if (require.main === module) {
    main().catch(console.error);
}
